# Data Loading and Analysis
*****************************
Loading, Preview, Sorting, Ranking, Filtering, Grouping and Aggregating in DataFrame.

### 1. Loading a Tabular Data File

Make sure the "titanic.csv" file (provided as part of this excercise) is located in the JupyterLab working directory. <br> Read the "titanic.csv" file dataset into a dataframe called df. Print the first 5 rows in df.

In [1]:
import pandas as pd
df = pd.read_csv("titanic.csv")
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

Read again the "titanic.csv" file dataset into a dataframe called df but without the following columns: "SibSp", "Parch" and "Ticket". Print the first 5 rows in df.

In [2]:
df.drop(['SibSp','Parch','Ticket'],axis=1).head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age     Fare  \
0                            Braund, Mr. Owen Harris    male  22.0   7.2500   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0  71.2833   
2                             Heikkinen, Miss. Laina  female  26.0   7.9250   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0  53.1000   
4                           Allen, Mr. William Henry    male  35.0   8.0500   

  Cabin Embarked  
0   NaN        S  
1   C85        C  
2   NaN        S  
3  C123        S  
4   NaN        S

Print the data type of each column by using the dtypes attribute.

In [3]:
df.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

Read again the "titanic.csv" file dataset into a dataframe called df but without the following columns: "SibSp", "Parch" and "Ticket" and also convert the column "Survived" into a boolean data type. Print the first 5 rows in df.

In [4]:
df['Survived']= df['Survived'].astype(bool)
df.drop(['SibSp','Parch','Ticket'],axis=1,inplace=True)
df.head()

PassengerId  Survived  Pclass  \
0            1     False       3   
1            2      True       1   
2            3      True       3   
3            4      True       1   
4            5     False       3   

                                                Name     Sex   Age     Fare  \
0                            Braund, Mr. Owen Harris    male  22.0   7.2500   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0  71.2833   
2                             Heikkinen, Miss. Laina  female  26.0   7.9250   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0  53.1000   
4                           Allen, Mr. William Henry    male  35.0   8.0500   

  Cabin Embarked  
0   NaN        S  
1   C85        C  
2   NaN        S  
3  C123        S  
4   NaN        S

Print the data type of each column by using the dtypes attribute. The "Survived" should be boolean data type.

In [5]:
df.dtypes

PassengerId      int64
Survived          bool
Pclass           int64
Name            object
Sex             object
Age            float64
Fare           float64
Cabin           object
Embarked        object
dtype: object

### 2. Preview the DataFrame

Print the two index components of the df dataframe: index, columns.

In [6]:
print(df.index,df.columns)

RangeIndex(start=0, stop=891, step=1) Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'Fare',
       'Cabin', 'Embarked'],
      dtype='object')


Print the size of the dataframe using the shape attribute. We have 891 rows and 9 columns. 

In [7]:
df.shape

(891, 9)

Print the amount of rows per each column without missing values.

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    bool   
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   Fare         891 non-null    float64
 7   Cabin        204 non-null    object 
 8   Embarked     889 non-null    object 
dtypes: bool(1), float64(2), int64(2), object(4)
memory usage: 56.7+ KB


Print the unique values for the "Sex" column.

In [9]:
df['Sex'].unique()

array(['male', 'female'], dtype=object)

Print the unique values for the "Embarked" column.

In [10]:
df['Embarked'].unique()

array(['S', 'C', 'Q', nan], dtype=object)

Print the amount of unique values for the "Embarked" column.

In [11]:
df['Embarked'].nunique()

3

Print the number of rows with the same specific value in the "Embarked" column.

In [12]:
df['Embarked'].value_counts()

S    644
C    168
Q     77
Name: Embarked, dtype: int64

How many passengers survived?

In [13]:
df['Survived'].value_counts()[True]

342

What is the survival percentage?

In [14]:
df['Survived'].mean()*100

38.38383838383838

### 3. Using Summary Statistics

What is the maximum age of the titanic passengers?

In [15]:
df['Age'].max()

80.0

What is the average age and average ticket price?

In [16]:
df[['Age','Fare']].mean()

Age     29.699118
Fare    32.204208
dtype: float64

How many passengers survived? (using the sum method).

In [17]:
df[df['Survived']==True].sum()['Survived']

C:\Users\158372\AppData\Local\Temp\ipykernel_15492\2421521197.py:1: FutureWarning: The default value of numeric_only in DataFrame.sum is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  df[df['Survived']==True].sum()['Survived']


342

Print a full statistical summary using the describe method. What is the agev value that 25% of passengers are smaller then this age? (should be 20.123).

In [18]:
df.describe()

PassengerId      Pclass         Age        Fare
count   891.000000  891.000000  714.000000  891.000000
mean    446.000000    2.308642   29.699118   32.204208
std     257.353842    0.836071   14.526497   49.693429
min       1.000000    1.000000    0.420000    0.000000
25%     223.500000    2.000000   20.125000    7.910400
50%     446.000000    3.000000   28.000000   14.454200
75%     668.500000    3.000000   38.000000   31.000000
max     891.000000    3.000000   80.000000  512.329200

### 4. Sorting and Ranking

Sort all rows in the dataframe using the 'Age' column is descending order and print the first 5 rows. 

In [19]:
df.sort_values('Age',ascending=False).head()

PassengerId  Survived  Pclass                                  Name  \
630          631      True       1  Barkworth, Mr. Algernon Henry Wilson   
851          852     False       3                   Svensson, Mr. Johan   
493          494     False       1               Artagaveytia, Mr. Ramon   
96            97     False       1             Goldschmidt, Mr. George B   
116          117     False       3                  Connors, Mr. Patrick   

      Sex   Age     Fare Cabin Embarked  
630  male  80.0  30.0000   A23        S  
851  male  74.0   7.7750   NaN        S  
493  male  71.0  49.5042   NaN        C  
96   male  71.0  34.6542    A5        C  
116  male  70.5   7.7500   NaN        Q

Sort all rows in the dataframe using the 'Age' column is descending order and save it in the same dataframe. Print the first 5 rows. 

In [20]:
df.sort_values('Age',ascending=False,inplace=True)

Sort the data based on male and female groups (female group is first) and inside of each group, sort the rows by age in ascending order. Present just the following columns 'Sex', 'Name', 'Age'.

In [21]:
df.sort_values(['Sex','Age'],ascending=[True,True])[['Sex','Name','Age']]

Sex                           Name   Age
644  female         Baclini, Miss. Eugenie  0.75
469  female  Baclini, Miss. Helene Barbara  0.75
381  female    Nakid, Miss. Maria ("Mary")  1.00
172  female   Johnson, Miss. Eleanor Ileen  1.00
642  female  Skoog, Miss. Margit Elizabeth  2.00
..      ...                            ...   ...
839    male           Marechal, Mr. Pierre   NaN
846    male       Sage, Mr. Douglas Bullen   NaN
859    male               Razi, Mr. Raihed   NaN
868    male    van Melkebeke, Mr. Philemon   NaN
878    male             Laleff, Mr. Kristo   NaN

[891 rows x 3 columns]

Drop the missing values and then sort the data based on male and female groups (female group is first) and inside of each group, sort the rows by age in ascending order. Present just the following columns 'Sex', 'Name', 'Age'.

In [22]:
df.dropna(inplace=True)
df.sort_values(['Sex','Age'],ascending=[True,True])[['Sex','Name','Age']]

Sex                                  Name   Age
205  female            Strom, Miss. Telma Matilda   2.0
297  female          Allison, Miss. Helen Loraine   2.0
10   female       Sandstrom, Miss. Marguerite Rut   4.0
618  female           Becker, Miss. Marion Louise   4.0
435  female             Carter, Miss. Lucile Polk  14.0
..      ...                                   ...   ...
54     male        Ostby, Mr. Engelhart Cornelius  65.0
456    male             Millet, Mr. Francis Davis  65.0
745    male          Crosby, Capt. Edward Gifford  70.0
96     male             Goldschmidt, Mr. George B  71.0
630    male  Barkworth, Mr. Algernon Henry Wilson  80.0

[183 rows x 3 columns]

What are the five oldest passengers that paid a ticket that cost more than the average ticket price? (create a boolean condition and than use it to filter the dataframe and than use the nlargest.

In [23]:
df[df['Fare']>df['Fare'].mean()].sort_values('Age',ascending=False).head()

PassengerId  Survived  Pclass  \
438          439     False       1   
587          588      True       1   
268          269      True       1   
195          196      True       1   
659          660     False       1   

                                              Name     Sex   Age      Fare  \
438                              Fortune, Mr. Mark    male  64.0  263.0000   
587               Frolicher-Stehli, Mr. Maxmillian    male  60.0   79.2000   
268  Graham, Mrs. William Thompson (Edith Junkins)  female  58.0  153.4625   
195                           Lurette, Miss. Elise  female  58.0  146.5208   
659                     Newell, Mr. Arthur Webster    male  58.0  113.2750   

           Cabin Embarked  
438  C23 C25 C27        S  
587          B41        C  
268         C125        S  
195          B80        C  
659          D48        C

### 5. Filtering

Filter and keep only the male passengers.

In [24]:
df[df['Sex']=='male']

PassengerId  Survived  Pclass                                  Name  \
630          631      True       1  Barkworth, Mr. Algernon Henry Wilson   
96            97     False       1             Goldschmidt, Mr. George B   
745          746     False       1          Crosby, Capt. Edward Gifford   
54            55     False       1        Ostby, Mr. Engelhart Cornelius   
456          457     False       1             Millet, Mr. Francis Davis   
..           ...       ...     ...                                   ...   
445          446      True       1             Dodge, Master. Washington   
193          194      True       2            Navratil, Master. Michel M   
340          341      True       2        Navratil, Master. Edmond Roger   
183          184      True       2             Becker, Master. Richard F   
305          306      True       1        Allison, Master. Hudson Trevor   

      Sex    Age      Fare    Cabin Embarked  
630  male  80.00   30.0000      A23        S  
96   male  71.00   34.6542       A5        C  
745  male  70.00   71.0000      B22        S  
54   male  65.00   61.9792      B30        C  
456  male  65.00   26.5500      E38        S  
..    ...    ...       ...      ...      ...  
445  male   4.00   81.8583      A34        S  
193  male   3.00   26.0000       F2        S  
340  male   2.00   26.0000       F2        S  
183  male   1.00   39.0000       F4        S  
305  male   0.92  151.5500  C22 C26        S  

[95 rows x 9 columns]

Filer out all passengers below the age of 40 and that are also 'male'. Print the ten first results.

In [25]:
df[(df['Sex'] == 'male') & (df['Age'] < 40)].head(10)

PassengerId  Survived  Pclass                                      Name  \
806          807     False       1                    Andrews, Mr. Thomas Jr   
332          333     False       1                 Graham, Mr. George Edward   
224          225      True       1              Hoyt, Mr. Frederick Maxfield   
137          138     False       1               Futrelle, Mr. Jacques Heath   
273          274     False       1                     Natsch, Mr. Charles H   
248          249      True       1             Beckwith, Mr. Richard Leonard   
148          149     False       2  Navratil, Mr. Michel ("Louis M Hoffman")   
679          680      True       1        Cardeza, Mr. Thomas Drake Martinez   
390          391      True       1                Carter, Mr. William Ernest   
512          513      True       1                 McGough, Mr. James Robert   

      Sex   Age      Fare        Cabin Embarked  
806  male  39.0    0.0000          A36        S  
332  male  38.0  153.4625          C91        S  
224  male  38.0   90.0000          C93        S  
137  male  37.0   53.1000         C123        S  
273  male  37.0   29.7000         C118        C  
248  male  37.0   52.5542          D35        S  
148  male  36.5   26.0000           F2        S  
679  male  36.0  512.3292  B51 B53 B55        C  
390  male  36.0  120.0000      B96 B98        S  
512  male  36.0   26.2875          E25        S

### 6. Grouping and Aggregating

What is the average ticket price for male versus female passengers?

In [26]:
df[['Sex','Fare']].groupby('Sex').mean()

Fare
Sex              
female  89.000900
male    69.124343

What is the average age for the survived group?

In [27]:
 df[df['Survived']]['Age'].mean()

32.905853658536586

How many survived group by the embarked port?

In [28]:
df[['Embarked','Survived']].groupby('Embarked').sum()

Survived
Embarked          
C               48
Q                1
S               74

What is the average age as well as the average price for male and female?

In [29]:
df[['Age','Sex','Fare']].groupby('Sex').mean()

Age       Fare
Sex                         
female  32.676136  89.000900
male    38.451789  69.124343

### End of document
************************